compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java

```diff
@@ -294,8 +299,21 @@ BLangStatementExpression desugar(BLangQueryExpr queryExpr, SymbolEnv env,
             BType refType = Types.getReferredType(queryExpr.getBType());
             if (TypeTags.isXMLTypeTag(refType.tag) || (refType.tag == TypeTags.UNION
                     && ((BUnionType) refType).getMemberTypes().stream().allMatch(memType ->
-                    TypeTags.isXMLTypeTag(Types.getReferredType(memType).tag)))) {
-                result = getStreamFunctionVariableRef(queryBlock, QUERY_TO_XML_FUNCTION, Lists.of(streamRef), pos);
+                    TypeTags.isXMLTypeTag(Types.getReferredType(memType).tag == TypeTags.INTERSECTION ?
+                            ((BIntersectionType) Types.getReferredType(memType)).effectiveType.tag :
+                            Types.getReferredType(memType).tag)))) {
+                if (refType.tag == TypeTags.UNION) {
```

Comment:
Can we just do `types.isSubTypeOfReadOnly(refType, env)`?

Reply:
It is private method. Is that okay to change it to public. If so I think we can you this method.

Reply:
that should be fine

Reply:
Added in https://github.com/ballerina-platform/ballerina-lang/pull/36576/commits/eac8c96aa024c308022b55c059e78532f19b30b9.